core/tests/behavior/write.rs

```diff
@@ -1233,3 +1234,31 @@ pub async fn test_fuzz_unsized_writer(op: Operator) -> Result<()> {
     op.delete(&path).await.expect("delete must succeed");
     Ok(())
 }
+
+/// seeking a negative position should return a InvalidInput error
+pub async fn test_invalid_reader_seek(op: Operator) -> Result<()> {
+    let path = uuid::Uuid::new_v4().to_string();
+    debug!("Generate a random file: {}", &path);
+    let (content, _) = gen_bytes();
+
+    op.write(&path, content.clone())
+        .await
+        .expect("write must succeed");
+
+    let mut r = op.reader(&path).await?;
+    let res = r.seek(std::io::SeekFrom::Current(-1024)).await;
+
+    assert!(res.is_err());
+
+    // the res should be a std::io::Error, which contains a ErrorKind::InvalidInput
```

Comment:
Oh, that's not I expected. Our users should don't depend on downcast to get the correct error kind. There are some extra work we need to:

https://github.com/apache/incubator-opendal/blob/58226a619da06f9b018955840ff976377a4a86b1/core/src/types/error.rs#L345-L355

We should mapping the opendal's error kind to io::ErrorKind correctly.

Reply:
Oh, That why I felt strange all the time. Should I add a new PR or added in this?

Reply:
We can add in this PR directly

Reply:
I found that
 `let res = r.seek(std::io::SeekFrom::Current(-1024)).await;`
 will return a std error `Interrupted` not `InvalidInput`. It seems not a good idea to map them together.
 https://github.com/apache/incubator-opendal/blob/58226a619da06f9b018955840ff976377a4a86b1/core/src/raw/oio/read.rs#L172
 
 I am not sure how to do that here

Reply:
/cc @Xuanwo  for ping

Reply:
> It seems not a good idea to map them together.

I think we should implement an internal convert function that convert opendal::Error to io::Error (please don't implement `From<std::io::Error> for opendal::Error`, we don't want to leak this to user).

Reply:
```rs
    let err = res.unwrap_err();
    
    assert_eq!(_convert_to_error(err).kind(), ErrorKind::InvalidInput);

```
Do you means a internal convert function like that?
```rs
fn _convert_to_error(err: std::io::Error) -> Error {
    let kind = match err.kind() {
        std::io::ErrorKind::Interrupted => ErrorKind::InvalidInput,
        _ => ErrorKind::Unexpected
    };

    Error::new(kind, err.to_string().as_str())
}
```

Reply:
The problem happens here:

https://github.com/apache/incubator-opendal/blob/11eb50fb67bb35c203a6d1e009429c6a7afd751f/core/src/raw/oio/read.rs#L164-L174

We convert all `opendal::Error` into `std::io::Error` with `std::io::ErrorKind::Interrupted` directly which is wrong. Instead, we should implement a  `convert_to_io_error(err: opendal::Error) -> std::io::Error` so we can return the correct `std::io::Error` to our users.